h5py/_npystrings.pyx

```diff
@@ -0,0 +1,255 @@
+    cdef const char* buf = info[0].contig[info[0].i]
+    # Deep copy char* from h5py into the numpy array
+    res = NpyString_pack(
+        info[0].allocator,
+        <npy_packed_static_string*>elem,
+        buf,
+        strlen(buf),
+    )
+    info[0].i += 1
+    return res
+
+
+def vstrings_scatter(hid_t space, size_t contig, size_t noncontig, size_t descr):
+    """Pure-python API interface to _proxy.pyx. This is necessary to
+    allow this module to be conditionally imported, allowing numpy 1.x to
+    continue working everywhere else.
+    """
+    _vstrings_scatter(space, <void *>contig, <void *>noncontig, <PyArray_Descr *>descr)
+
+
+cdef void _vstrings_scatter(hid_t space, void *contig, void *noncontig,
+                            PyArray_Descr *descr):
+    """Convert a zero-terminated char**, which is the in-memory representation
+    for a HDF5 variable-width string dataset, into a NpyString array
+    (NumPy's native variable-width strings dtype).
+
+    Parameters
+    ----------
+    space : hid_t
+        HDF5 dataspace ID for the output non-contiguous buffer
+        containing packed NpyStrings
+    contig : void * (actually a char **)
+        Input buffer for the contiguous zero-terminated char **
+    noncontig : void *
+        Output buffer for the non-contiguous packed NpyStrings
+    descr : PyArray_Descr * (actually a PyArray_StringDTypeObject *)
+        NumPy dtype descriptor for the NpyStrings
+    npoints : size_t
+        Number of points to copy
+
+    Notes
+    -----
+    This function deep-copies the input zero-terminated strings.
+    Memory management for the outputs is handled by NumPy.
+    """
+    cdef vstrings_scatter_t info
+    info.i = 0
+    info.contig = <const char**>contig
+    info.allocator = NpyString_acquire_allocator(<PyArray_StringDTypeObject *>descr)
+    if info.allocator is NULL:
+        raise RuntimeError("Failed to acquire string allocator")
+
+    # H5Diterate needs a tid of the correct size.
+    # Disregard mtype from _proxy::dset_rw.
+    # The memory type is actually npy_packed_static_string
+    # (an opaque struct of arbitrary size)
+    # and not HDF5 variable length strings (char*).
+    tid = H5Tcreate(H5T_OPAQUE, SIZEOF_NPY_PACKED_STATIC_STRING)
+
+    # Read char*[] (zero-terminated) from h5py
+    # and deep-copy to npy_packed_static_string[] for numpy
+    H5Diterate(noncontig, tid, space, vstrings_scatter_cb, &info)
+    NpyString_release_allocator(info.allocator)
+    H5Tclose(tid)
+
+
+cdef herr_t vstrings_gather_cb(
+    void *elem, hid_t type_id, unsigned ndim,
+    const hsize_t *point, void *operator_data
+) except -1:
+    cdef vstrings_gather_t *info = <vstrings_gather_t *>operator_data
+    cdef npy_static_string *unpacked = info[0].unpacked + info[0].i
+    # Obtain a reference to the string (NOT zero-terminated) and its size
+    res = NpyString_load(info[0].allocator, <npy_packed_static_string*>elem, unpacked)
+    info[0].i += 1
+    return res
```

Comment:
The NEP says that NpyString_load is allowed to return 1 to indicate a null string, and the [H5Diterate docs](https://support.hdfgroup.org/documentation/hdf5/latest/group___h5_d.html#ga71421c684884ab49765748720fe938db) imply that returning a non-zero value exits the loop, although this isn't actually specified. I couldn't trigger a problem in a brief test, but do we need any special handling around this?

Reply:
Couldn't reproduce it either. For the sake of caution, I've added special handling for it and added a unit test.